### ISSUE TITLE
FillTriangle using Canvas does not work correctly.

### ISSUE THREAD

**Issue by KurtE** [adafruit/Adafruit-GFX-Library#325]
There is a forum thread up on PJRC forum:
https://forum.pjrc.com/threads/64589-fillTriange-issue-with-screen-or-buffer-size

Where a user had issues with a fillRectangle not working correctly on an ILI9341 display.   
It screws up two different ways, depending on the actual triangle.  The version they posted filled only part of the triangle.  If you change the one value (in the case that I post) the 310, it goes completely outside of the bounds of the triangle.

Note: It works fine if you draw not using the canvas.  I verified that with your library as well as my own ILI9341_t3n. 
Example sketch that shows the issue.
```
#define USE_CANVAS
//#include <ILI9341_t3n.h>
#include <Adafruit_GFX.h>
#include <Adafruit_ILI9341.h>
#include <SPI.h>
const uint16_t  tft_Color_Black        = 0x0000;
const uint16_t  tft_Color_Blue         = 0x001F;
const uint16_t  tft_Color_Red          = 0xF800;

#define TFT_CS 10
#define TFT_DC  9
#define TFT_RST  8

//ILI9341_t3n tft = ILI9341_t3n(TFT_CS, TFT_DC, TFT_RST);
Adafruit_ILI9341 tft = Adafruit_ILI9341(TFT_CS, TFT_DC, TFT_RST);

void setup() {

  delay(250);

  tft.begin();
  tft.setFont();
  tft.fillScreen(tft_Color_Black);
  tft.setRotation(0);

}

int loop_count = 0;
#ifdef USE_CANVAS
GFXcanvas16 *canvas = new GFXcanvas16 (240, 320);  //
#endif

void loop() {
  loop_count++;
#ifdef USE_CANVAS
  if (loop_count & 1) {
    canvas -> drawTriangle(0, 0, 240, 0, 120, 320, ILI9341_RED);  // This works fine
    canvas -> fillTriangle(0, 0, 240, 0, 120, 310, ILI9341_BLUE);  // This is what stops halfway down screen.
  } else { 
    canvas -> drawTriangle(0, 0, 240, 0, 120, 320, ILI9341_BLUE);  // This works fine
    canvas -> fillTriangle(0, 0, 240, 0, 120, 310, ILI9341_RED);  // This is what stops halfway down screen.
  }
  tft.drawRGBBitmap(0 , 0 , canvas -> getBuffer () , 240 , 320);
#else
  if (loop_count & 1) {
    //    tft.useFrameBuffer(0); // make sure frame buffer is turned off.
    tft.drawTriangle(0, 0, 240, 0, 120, 320, ILI9341_RED);  // This works fine
    tft.fillTriangle(0, 0, 240, 0, 120, 310, ILI9341_BLUE);  // This is what stops halfway down screen.
  } else {
    //   tft.useFrameBuffer(1);
    tft.drawTriangle(0, 0, 240, 0, 120, 320, ILI9341_BLUE);  // This works fine
    tft.fillTriangle(0, 0, 240, 0, 120, 310, ILI9341_RED);  // This is what stops halfway down screen.
    //    tft.updateScreen();
  }
#endif  
  delay(250);
}
```
If you comment out the USE_CANVAS you will see the main triangle code worked.  Note: I fixed something similar in my own library awhile ago where some variables needed to be changed from 16 bits to 32 bits.

**Comment by KurtE** [comment]
Found the issue:  I am not sure how many other places this may hit:  But the code funnels down to the method: 
```
void GFXcanvas16::drawFastRawHLine(int16_t x, int16_t y, int16_t w,
                                   uint16_t color) {
  // x & y already in raw (rotation 0) coordinates, no need to transform.
  size_t buffer_index = y * WIDTH + x;
  for (int16_t i = buffer_index; i < buffer_index + w; i++) {
    buffer[i] = color;
  }
}
```
Now suppose you call this with: x=105, y=273 w=31... 
This computes the starting buffer index of 65625 which will not fit into an int16_t which the for loop has for the index code...

Again not sure if there are others.  But changing this to int32_t or uint32_t solved the problem.  I actually changed that as well as the size_t of buffer_index to uint32_t and removed a compiler warning I was getting.
